File: 백준/Gold/17142. 연구소 3/연구소 3.js
const fs = require('fs');
const input = fs.readFileSync(0).toString().trim().split('\n');

const [N, M] = input[0].split(' ').map(Number);
const board = input.slice(1, N + 1).map(line => line.trim().split(' ').map(Number));

const dr = [-1, 1, 0, 0];
const dc = [0, 0, -1, 1];

function isInBound(r, c, n) {
    return r >= 0 && r < n && c >= 0 && c < n;
}

function combination(arr, n) {
    if (n === 1) return arr.map(value => [value]);

    const results = [];
    arr.forEach((fixed, index) => {
        const rest = arr.slice(index + 1);
        const combis = combination(rest, n - 1);
        const combine = combis.map(combi => [fixed, ...combi]);
        results.push(...combine);
    });
    return results;
}

class Queue {
    constructor() {
        this.queue = [];
        this.head = 0;
        this.tail = 0;
    }
    push(value) {
        this.queue[this.tail++] = value;
    }
    pop() {
        return this.queue[this.head++];
    }
    isEmpty() {
        return this.head === this.tail;
    }
}

function solution() {
    const viruses = [];
    let totalEmpty = 0;

    for (let r = 0; r < N; r++) {
        for (let c = 0; c < N; c++) {
            if (board[r][c] === 2) viruses.push([r, c]);
            if (board[r][c] === 0) totalEmpty++;
        }
    }

    if (totalEmpty === 0) return 0;

    const virusCombinations = combination(viruses, M);
    let answer = Infinity;

    virusCombinations.forEach(virusPositions => {
        const visited = Array.from({ length: N }, () => Array(N).fill(-1));
        const queue = new Queue();
        let infectedEmpty = 0;
        let localMaxTime = 0;

        virusPositions.forEach(([r, c]) => {
            visited[r][c] = 0;
            queue.push([r, c, 0]);
        });

        while (!queue.isEmpty()) {
            const [r, c, time] = queue.pop();

            for (let i = 0; i < 4; i++) {
                const nr = r + dr[i];
                const nc = c + dc[i];

                if (isInBound(nr, nc, N) && visited[nr][nc] === -1 && board[nr][nc] !== 1) {
                    visited[nr][nc] = time + 1;
                
                    if (board[nr][nc] === 0) {
                        infectedEmpty++;
                        localMaxTime = time + 1;
                    }
                
                    queue.push([nr, nc, time + 1]);
                }
            }
        }

        if (infectedEmpty === totalEmpty) {
            answer = Math.min(answer, localMaxTime);
        }
    });

    return answer === Infinity ? -1 : answer;
}

console.log(solution());

File: 백준/Gold/17142. 연구소 3/연구소 3.py
import sys
from collections import deque
from itertools import combinations

def solution():
    input = sys.stdin.read().splitlines()
    if not input:
        return
    
    n, m = map(int, input[0].split())
    board = [list(map(int, line.split())) for line in input[1:n+1]]

    dr = [-1, 1, 0, 0]
    dc = [0, 0, -1, 1]

    viruses = []
    total_empty = 0

    for r in range(n):
        for c in range(n):
            if board[r][c] == 2:
                viruses.append((r, c))
            elif board[r][c] == 0:
                total_empty += 1

    if total_empty == 0:
        return 0

    virus_combinations = list(combinations(viruses, m))
    answer = float('inf')

    for combi in virus_combinations:
        visited = [[False] * n for _ in range(n)]
        queue = deque()
        
        infected_empty = 0
        local_max_time = 0

        for vr, vc in combi:
            visited[vr][vc] = True
            queue.append((vr, vc, 0))

        while queue:
            r, c, time = queue.popleft()

            for i in range(4):
                nr, nc = r + dr[i], c + dc[i]

                if 0 <= nr < n and 0 <= nc < n:
                    if not visited[nr][nc] and board[nr][nc] != 1:
                        visited[nr][nc] = True
                        
                        if board[nr][nc] == 0:
                            infected_empty += 1
                            local_max_time = time + 1
                        
                        queue.append((nr, nc, time + 1))
        
        if infected_empty == total_empty:
            answer = min(answer, local_max_time)

    return answer if answer != float('inf') else -1

print(solution())